api/src/main/java/org/openmrs/module/biologicalorder/OrderResultEventListener.java

```diff
@@ -85,158 +88,102 @@ public void onMessage(Message message) {
-			Encounter encounter = new Encounter();
-			encounter.setForm(form);
-			encounter.setPatient(patient);
-			encounter.setEncounterType(encounterType);
-			encounter.setEncounterDatetime(obs.getObsDatetime());
-			encounter.setLocation(defaultLocation);
-			encounter.setProvider(encounterRole, provider);
-			Encounter encounterSaved = Context.getEncounterService().saveEncounter(encounter);
-			
-			obs2.setEncounter(encounterSaved);
-			obs3.setEncounter(encounterSaved);
-			if (isDoubleValue) {
-				obs4.setEncounter(encounterSaved);
-			}
-			obs5.setEncounter(encounterSaved);
-			obs6.setEncounter(encounterSaved);
-			obs7.setEncounter(encounterSaved);
-			
-			obsService.saveObs(obs2, userUuid);
-			obsService.saveObs(obs3, userUuid);
+
+			Obs accessionNumber = createObs(getConceptByUuid("162086AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"));
+			accessionNumber.setValueText(obs.getEncounter().getOrders().iterator().next().getAccessionNumber());
+
+			Obs grossViralLoad = createObs(getConceptByUuid("CI0050051AAAAAAAAAAAAAAAAAAAAAAAAAAA"));
+			grossViralLoad.setValueText(obs.getValueText());
+
+			Obs releasedDate = createObs(getConceptByUuid("CI0050052AAAAAAAAAAAAAAAAAAAAAAAAAAA"));
+			releasedDate.setValueDatetime(obs.getObsDatetime());
+
+			Encounter encounterSaved = createEncounter();
+			viralLoad.setEncounter(encounterSaved);
+			viralLoadQualitative.setEncounter(encounterSaved);
+			accessionNumber.setEncounter(encounterSaved);
+			grossViralLoad.setEncounter(encounterSaved);
+			releasedDate.setEncounter(encounterSaved);
+			
+			obsService.saveObs(viralLoad, userUuid);
+			obsService.saveObs(viralLoadQualitative, userUuid);
+			obsService.saveObs(accessionNumber, userUuid);
+			obsService.saveObs(grossViralLoad, userUuid);
+			obsService.saveObs(releasedDate, userUuid);
 			if (isDoubleValue) {
-				obsService.saveObs(obs4, userUuid);
+				viralLoadLog.setEncounter(encounterSaved);
+				obsService.saveObs(viralLoadLog, userUuid);
 			}
-			obsService.saveObs(obs5, userUuid);
-			obsService.saveObs(obs6, userUuid);
-			obsService.saveObs(obs7, userUuid);
-			
-			System.out.println("********** END PROCESS MESSAGE ****************");
 		}
 	}
-	
-	public Obs getLastBiologicalObs(Integer personId) throws APIException {
-		String hql = "SELECT * FROM obs o JOIN encounter e ON e.encounter_id = o.encounter_id WHERE e.encounter_type = 8 AND DATE(o.date_created) = CURDATE() AND o.person_id = "
-		        + personId;
-		return (Obs) getSession().createQuery(hql).uniqueResult();
+
+	private static Concept getConceptByUuid(String conceptUUID) {
+		return Context.getConceptService().getConceptByUuid(conceptUUID);
+	}
+
+	private Encounter createEncounter() {
+		EncounterType encounterType = new EncounterType(ENCOUNTER_TYPE_ID);
+		Form form = Context.getFormService().getForm(FORM_ID);
+		EncounterRole encounterRole = Context.getEncounterService().getEncounterRole(2);
+		Provider provider = Context.getProviderService().getProvider(1);
```

Comment:
1. These should be UUIDs instead of IDs.
2. The UUIDs should be passed in as parameters to the function 
`
private Encounter createEncounter(String encTypeUUID, String formUUID, String encRoleUUID){}
`
3. The provider should Ideally be the currently active provider unless otherwise.